### DISCUSSION TITLE
Arroost programming: memory, XOR, AND → bit adder

### DISCUSSION THREAD

**Discussion by anderium** [TodePond/Arroost#297] [Show and tell]
I've been able to create some important behaviours for programming with Arroost.  They're quite minor so far, but theoretically it should already be enough to do everything.  Namely, I've created a reality without the arrow of reality, because they're bugged (see #295).  I've then created an `XOR` gate and an `AND` gate.  Initially I based these on a rising edge and falling edge detector, but I was able to simplify it away.  I'll probably publish those later.  The most important thing to know about all these coding gates I've made, is that they use green wire to transfer the signal, this allows the signal to be blocked and let through with a single activation of the arrow of colour.

## Memory cell

Anyway, first up, the memory cell:  It takes a single pulse as input, and toggles a continuous pulse on/off.  This is exactly what a reality is.  It does this with a simple cycle that gets started and interrupted with each next input pulse.

![arroost_memory_cell](https://github.com/TodePond/Arroost/assets/33520919/d5e80e5f-c29b-46b8-89f4-d38d1bbc6f51)

## AND gate

Next is the `AND` gate:  It takes two inputs and gives one output only if both of the inputs are active.  This is easy enough if the pulse is synced, but if you have a continuous pulse—let's call that a _voltage_—the delay when changing colours becomes an issue.  To work around this, I create a voltage splitter that separates the input voltage into two paths.  (The same path should be taken by both inputs, so it's even a synchronised splitter!)  This allows functions to make two cycle computations instead of only one cycles, this gives them enough time to reset themselves.  The `AND` gate blocks the voltage of one of its inputs by default, and only lets it pass when the other input is also on.

![arroost_AND_gate](https://github.com/TodePond/Arroost/assets/33520919/a4ab08d1-65c5-4191-9fe1-44cc0c49120e)

## XOR gate

Finally the `XOR` gate:  It also takes two inputs and gives one output, but only if exactly one of the inputs is active.  This is more complicated than the `AND` gate, but very similar in concept.  Instead of one input opening a path for the other voltage, it has to close the path for that one.  Additionally, that voltage has to block the path of the first.  This way, if both voltages are on they block each other, but if only one of them is on it can pass through (while blocking the nothingness of the voltage that's off).

![arroost_XOR_gate](https://github.com/TodePond/Arroost/assets/33520919/6931d647-4e8d-4cc9-b94d-818366dbcbfd)

## Bit adder

A single bit adder is actually very simple to make with the building blocks above.  The low bit is the `XOR` of the two input bits and the high bit is the `AND` of the inputs.

![arroost_bit_adder](https://github.com/TodePond/Arroost/assets/33520919/6823714e-0e9c-45a0-8786-452077f14690)

## Additional improvements

- To slightly declutter the circuits you can double-up arrows with no-delay and one-delay timings.
- The `AND` gate could be interpreted as `A if B` and the XOR gate as `(A unless B) OR (B unless A)`, but there's another way to define the `XOR` in terms of `AND` that uses less wire in this case: `(A OR B) unless (A AND B)`.  This is shorter because we have to calculate the `AND` anyway.  The downside is that it takes another cycle to get the output.

**Comment by TodePond** [comment]
This is incredible! 
it's it ok if i share these gifs on mastodon?

**Reply by anderium** [reply]
Yes feel free to!  I wasn't able to control the quality of gifs because my `ffmpeg` install is broken (still using Mac 10.14 Mojave), so I've attached the vids here in case you want the last one to be readable :D

<details>
<summary>Videos instead of gifs</summary>


https://github.com/TodePond/Arroost/assets/33520919/0cdf639a-5545-4645-84d9-20765935e859


https://github.com/TodePond/Arroost/assets/33520919/a7a1e590-e44d-4491-a501-38eb96fb828b


https://github.com/TodePond/Arroost/assets/33520919/940aa83a-d910-4f26-b792-f7b35cca4cb0


https://github.com/TodePond/Arroost/assets/33520919/41c18010-9329-47c6-922b-f52605f4e1b9


</details>

**Comment by alifeee** [comment]
this is wonderful!! your mind is grand 
how long til aroost within aroost /jk
